discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::str::FromStr;
+
+use async_trait::async_trait;
+use displaydoc::Display;
+use sqlx::{
+    sqlite::{Sqlite, SqliteConnectOptions, SqlitePoolOptions},
+    Pool,
+};
+use thiserror::Error;
+use url::Url;
+use uuid::Uuid;
+use xayn_discovery_engine_ai::GenericError;
+
+use crate::document::{HistoricDocument, Id};
+
+use super::Storage;
+
+#[derive(Error, Debug, Display)]
+pub enum Error {
+    /// Failed to initialize database: {0}
+    Init(#[source] sqlx::Error),
+
+    /// Failed to acquire connection: {0}
+    AcquireConnection(#[source] sqlx::Error),
+
+    /// Failed to migrate database: {0}
+    Migration(#[from] sqlx::migrate::MigrateError),
+
+    /// Failed to fetch: {0}
+    Fetch(#[source] sqlx::Error),
+
+    /// Failed to covert type: {0}
+    TypeConversion(#[source] GenericError),
+}
+
+#[derive(Clone)]
+pub(crate) struct SqliteStorage {
+    pool: Pool<Sqlite>,
+}
+
+impl SqliteStorage {
+    pub(crate) async fn connect(uri: &str) -> Result<Self, Error> {
+        let opt = SqliteConnectOptions::from_str(uri)
+            .map_err(Error::Init)?
+            .create_if_missing(true);
+
+        let pool = SqlitePoolOptions::new()
+            .connect_with(opt)
+            .await
+            .map_err(Error::Init)?;
+
+        Ok(Self { pool })
+    }
+}
+
+#[async_trait]
+impl Storage for SqliteStorage {
+    type StorageError = Error;
+
+    async fn init_database(&self) -> Result<(), <Self as Storage>::StorageError> {
+        sqlx::migrate!("src/storage/migrations")
+            .run(&self.pool)
+            .await
+            .map_err(Into::into)
+    }
+
+    async fn fetch_history(
+        &self,
+    ) -> Result<Vec<HistoricDocument>, <Self as Storage>::StorageError> {
+        #[derive(sqlx::FromRow)]
+        struct _HistoricDocument {
+            document: Uuid,
+            url: String,
+            snippet: String,
+            title: String,
+        }
+
+        let mut con = self
+            .pool
+            .acquire()
+            .await
+            .map_err(Error::AcquireConnection)?;
+
+        sqlx::query_as::<_, _HistoricDocument>("SELECT nr.document, nr.url, nr.snippet, nr.title FROM HistoricDocument AS hd, NewsResource AS nr ON hd.document = nr.document;")
```

Comment:
❓ the query tries to fetch the historic documents, but the sql string selects the news resource fields (with the same id), is this on purpose? i guess this is due to the way our sql tables are defined?

Reply:
yes exactly the `HistoricDocument` only has an id column. the idea is if we want to add/remove data then we only have to adjust the sql statement but not the tables. for example: let's say we also want to return the embedding then we just change the statement to 
`SELECT e.embedding, nr.document, nr.url, nr.snippet, nr.title FROM HistoricDocument AS hd, NewsResource AS nr, Embedding AS e ON hd.document = nr.document, hd.document = e.document;`